### PR TITLE
otelhttp: client metrics

### PR DESCRIPTION
**Pull request by Sovietaced**
This pull request attempts to follow up on the work that @rangelReale did in https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3769 and close out the work. This pull request adds support for http client metrics that mostly adhere to the [v1.20 semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/metrics/semantic_conventions/http-metrics.md#http-client). In many ways this code attempts to mirror the http server metric code in implementation. 

As such, it reuses the HTTP request body wrapper logic and introduces HTTP response body wrapper logic with callbacks to record metrics once the response body has been read.

Open Questions: 
* As written these client metrics follow the convention of the server metrics, but don't actually follow the convention of the OTEL spec as noted in https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4501. I guess this migration can be done later?
* I don't believe the code will report HTTP response metrics unless the response body is read, which is the responsibility of the client. Are we ok with this behavior? I tried to follow the the pattern where reading bodies is done with wrapping to minimize allocations. 
* It is not clear to me how the body wrappers work with long lived connections?

Fixes https://github.com/open-telemetry/opentelemetry-go-contrib/issues/3134
Fixes https://github.com/open-telemetry/opentelemetry-go-contrib/issues/1336

### CODE REVIEW

File: instrumentation/net/http/otelhttp/common.go
Comment (ash2k) [inline instrumentation/net/http/otelhttp/common.go:37]: I wonder if these constants need to be public? Technically renaming them is a breaking change. I think the smaller the exposed API surface it, the better it is. WDYT?

File: instrumentation/net/http/otelhttp/internal/semconvutil/httpconv.go
Comment (ash2k) [inline instrumentation/net/http/otelhttp/internal/semconvutil/httpconv.go:321]: This could be a single append call.

File: instrumentation/net/http/otelhttp/transport.go
Comment (ash2k) [inline instrumentation/net/http/otelhttp/transport.go:19]: I wonder if this project has a formatting check in CI...

File: instrumentation/net/http/otelhttp/transport.go
Comment (ash2k) [inline instrumentation/net/http/otelhttp/transport.go:196]: https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4590

File: instrumentation/net/http/otelhttp/transport_test.go
Comment (ash2k) [inline instrumentation/net/http/otelhttp/transport_test.go:25]: Inconsistent grouping here too.

**Review by ash2k — CHANGES_REQUESTED** [review]

File: instrumentation/net/http/otelhttp/transport.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go:196]: I agree with the general desire for this codebase to follow the OTEL specification, but I also see some value in following existing convention and then having a more thought out approach to migrating towards seconds across the entire codebase. 

This is the same reason I stuck with the incorrectly named metrics, so that this code would help reach feature parity with service side metrics

File: instrumentation/net/http/otelhttp/transport.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go:19]: It might be, I did most of this work on a plane with a 3 kb/s connection so unfortunately I wasn't able to download `golangci-lint` and reproduce the issues locally, I'll try and sort it out

File: instrumentation/net/http/otelhttp/internal/semconvutil/httpconv.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/internal/semconvutil/httpconv.go:321]: Good point

File: instrumentation/net/http/otelhttp/common.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/common.go:37]: Oh that is a good point, I forget that other go modules could be relying on these public constants. I don't see why they need to be public though. I'm only familiar with the gRPC metrics and those definitely don't follow this pattern.

File: instrumentation/net/http/otelhttp/transport_test.go
Comment (MadVikingGod) [inline instrumentation/net/http/otelhttp/transport_test.go:42]: Whichever test uses the SDK needs to be in the `test` directory.  To prevent this from depending on the SDK.

File: instrumentation/net/http/otelhttp/transport.go
Comment (ash2k) [reply instrumentation/net/http/otelhttp/transport.go:197]: Thank you. I'll leave this up to maintainers to decide.

File: instrumentation/net/http/otelhttp/common.go
Comment (ash2k) [reply instrumentation/net/http/otelhttp/common.go:37]: Another data point here - this module (`go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp`) is at `v0.46.1` so making breaking changes is still allowed. So we better get it in shape before it reaches 1.0. This includes hiding/removing everything that doesn't need to be exposed, renaming metrics/method/etc, changing units, etc. So now might be a good time as any to make these private, while it's not too late.

File: instrumentation/net/http/otelhttp/transport_test.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport_test.go]: Fixed in 1a4c8cb51c3ecbc82a27bbe897f74e545b051252

File: instrumentation/net/http/otelhttp/common.go
Comment (lenta-platform) [reply instrumentation/net/http/otelhttp/common.go:37]: At the moment there is global instrumentration vs semconv naming mismath, e.g. "http.client.duration" vs "http.client.request.duration", "http.server.duration" vs "http.server.request.duration" and so on.

Spec: https://opentelemetry.io/docs/specs/semconv/http/http-metrics/#metric-httpclientrequestduration

Will we drop semconv, is it outdated?

**Comment by pellared** [comment]
> I don't believe the code will report HTTP response metrics unless the response body is read, which is the responsibility of the client. Are we ok with this behavior? I tried to follow the the pattern where reading bodies is done with wrapping to minimize allocations.

We are fine. From https://pkg.go.dev/net/http:

> The caller must close the response body when finished with it:
> ...
> Caller should close resp.Body when done reading from it.

Also we already have this requirement for trace instrumentation.

Side note: probably we should call it out in package documentation.

**Comment by pellared** [comment]
> As written these client metrics follow the convention of the server metrics, but don't actually follow the convention of the OTEL spec as noted in https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4501. I guess this migration can be done later?

We want to first be be compliant with the [v1.20 semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/metrics/semantic_conventions/http-metrics.md#http-client).  _Next semantic convention versions have many breaking changes and we want to help a little with the migration._

File: instrumentation/net/http/otelhttp/common.go
Comment (pellared) [reply instrumentation/net/http/otelhttp/common.go:37]: These constants should live in `semconv` packages. See: https://github.com/open-telemetry/opentelemetry-go/issues/4528

I would rather not export it in `otelhttp` as I see no big value and it would increase the API surface.

File: instrumentation/net/http/otelhttp/transport.go
Comment (pellared) [reply instrumentation/net/http/otelhttp/transport.go:197]: It should be `ms` like defiened in https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/metrics/semantic_conventions/http-metrics.md#http-client. Resolving.

File: instrumentation/net/http/otelhttp/common.go
Comment (pellared) [inline instrumentation/net/http/otelhttp/common.go:43]: These do not match https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/metrics/semantic_conventions/http-metrics.md#http-client

Can we also fix the metrics for HTTP server instrumentation?

File: instrumentation/net/http/otelhttp/transport.go
Comment (pellared) [inline instrumentation/net/http/otelhttp/transport.go:115]: nit: Can we use the descriptions from https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/metrics/semantic_conventions/http-metrics.md#http-client? Just I would just get rid of "(compressed)".

File: instrumentation/net/http/otelhttp/common.go
Comment (pellared) [inline instrumentation/net/http/otelhttp/common.go:34]: Could we update the `otelhttp` package documentation to include the information about the emited metrics? Something like https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/runtime

This can be addressed as a separate issue and PR.

**Comment by pellared** [comment]
Please add a changelog entry.

**Comment by Sovietaced** [comment]
> Please add a changelog entry.

Will do, I didn't add one since you seemed a bit picky about the changelog entries on my last PR :)

File: instrumentation/net/http/otelhttp/common.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/common.go:34]: I will defer that to a separate issue/PR

File: instrumentation/net/http/otelhttp/test/transport_test.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/test/transport_test.go]: 0c097201a95fe872a4b19ba097f095cdae4353a7

File: instrumentation/net/http/otelhttp/common.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/common.go]: 0c097201a95fe872a4b19ba097f095cdae4353a7

File: instrumentation/net/http/otelhttp/transport.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go]: 0c097201a95fe872a4b19ba097f095cdae4353a7

File: instrumentation/net/http/otelhttp/common.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/common.go]: https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4774

File: instrumentation/net/http/otelhttp/common.go
Comment (pellared) [inline instrumentation/net/http/otelhttp/common.go:43]: Can we unexport these constants and add a changelog entry in "Removed" section?

```
- Remove `RequestCount`, `RequestContentLength`, `ResponseContentLength`, `ServerLatency` constants from `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp`. (#4707)
```

**Review by pellared — DISMISSED** [review]

File: instrumentation/net/http/otelhttp/common.go
Comment (SodaDev) [inline instrumentation/net/http/otelhttp/common.go:43]: I guess it is not aligned with semconv from OTEL: https://opentelemetry.io/docs/specs/semconv/http/http-metrics/#http-client

File: instrumentation/net/http/otelhttp/internal/semconvutil/httpconv.go
Comment (SodaDev) [inline instrumentation/net/http/otelhttp/internal/semconvutil/httpconv.go:301]: Maybe it could be replaced with
semconv.HTTPClientAttributesFromHTTPRequest(req) ?

Comment (pellared) [reply instrumentation/net/http/otelhttp/internal/semconvutil/httpconv.go:301]: The `semconv` does not have such function.

Comment (SodaDev) [reply instrumentation/net/http/otelhttp/internal/semconvutil/httpconv.go:301]: yes, it was deprecated at version 12. It was replaced with httpconv.ClientRequest(req) from "go.opentelemetry.io/otel/semconv/v1.20.0/httpconv"

File: instrumentation/net/http/otelhttp/common.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/common.go]: 233049a5211e6aebcbdd90d6fd4fe4d50e114f49

Comment (pellared) [reply instrumentation/net/http/otelhttp/common.go]: Server* constans have to be unexported as well.

Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/common.go]: My bad, I misunderstood. Will take care of that when I get back to the computer.

**Review by pellared — APPROVED** [review]

**Comment by Sovietaced** [comment]
@dmathieu @Aneurysm9 @dashpole @hanyuancheung Please take a look when you have a chance. TY

File: instrumentation/google.golang.org/grpc/otelgrpc/example/api/hello-service.pb.go
Comment (dmathieu) [inline instrumentation/google.golang.org/grpc/otelgrpc/example/api/hello-service.pb.go:8]: Why are those new lines here? They seem unrelated to this PR.

Comment (Sovietaced) [reply instrumentation/google.golang.org/grpc/otelgrpc/example/api/hello-service.pb.go:8]: Hmm. Not sure. Let me back those out

Comment (Sovietaced) [reply instrumentation/google.golang.org/grpc/otelgrpc/example/api/hello-service.pb.go]: Fixed in f54fc6c7805342222601f1edfc27d94d642fc9a0

**Review by dmathieu — APPROVED** [review]

File: instrumentation/net/http/otelhttp/transport.go
Comment (dashpole) [inline instrumentation/net/http/otelhttp/transport.go:98]: Should this be `Measures the size of HTTP request messages (compressed).`?

https://github.com/open-telemetry/opentelemetry-specification/blob/855d80c93ffc11ee7c96177be8082989afb020f5/semantic_conventions/metrics/http.yaml#L56C13-L56C69

Comment (dashpole) [reply instrumentation/net/http/otelhttp/transport.go:98]: Same for response below, and for server request/response.

Comment (pellared) [reply instrumentation/net/http/otelhttp/transport.go:98]: I noticed that it was `(uncompressed)` before. Isn't `http` package handling the transport compression and therefore we get the uncompressed message size value?

Comment (dashpole) [reply instrumentation/net/http/otelhttp/transport.go:98]: Ah.  Makes sense.  We aren't really correctly implementing this metric then, are we?

Comment (pellared) [reply instrumentation/net/http/otelhttp/transport.go:98]: This is how I see it. I was not checking if it is possible to implement it correctly.

**Review by dashpole — APPROVED** [review]

**Comment by Sovietaced** [comment]
@pellared is there anything else blocking this from being merged?

**Comment by pellared** [comment]
@Sovietaced No blockers. I am just giving time for @MadVikingGod until end of week to review it.

File: instrumentation/net/http/otelhttp/transport.go
Comment (MadVikingGod) [inline instrumentation/net/http/otelhttp/transport.go:257]: Currently, this will only record on an EOF.  Is there a way to record on an EOF or a `close()` but only record once?

The use case I'm thinking of is if a client reads part-way and is canceled.

File: instrumentation/net/http/otelhttp/common.go
Comment (ash2k) [reply instrumentation/net/http/otelhttp/common.go:36]: `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` is at `v0.46.1`. If now is not the right moment to make breaking changes, then... when is? I may be wrong, but I thought breaking changes are ok and had happened in `v0.x.x` modules here before. 🤔

Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/common.go:36]: I originally tried not to include any breaking changes but got guidance from @pellared to update these: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4707#issuecomment-1871335266

File: instrumentation/net/http/otelhttp/transport.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go:257]: Good point, I will add a test case for that and update the code.

Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go]: Addressed in 0411674549f801f9aa02f2c91a46dc6967854a03. Let me know how that looks :)

File: instrumentation/net/http/otelhttp/common.go
Comment (pellared) [reply instrumentation/net/http/otelhttp/common.go:36]: @MadVikingGod

File: instrumentation/net/http/otelhttp/transport.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go]: ping @MadVikingGod

**Review by hanyuancheung — APPROVED** [review]

File: instrumentation/net/http/otelhttp/transport.go
Comment (MadVikingGod) [inline instrumentation/net/http/otelhttp/transport.go:165]: Why are we wrapping the body, with a recorder that does nothing?

File: instrumentation/net/http/otelhttp/transport.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go:194]: There was a comment about not using the content length in the previous pull request for this feature here: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3769#discussion_r1333577399

File: instrumentation/net/http/otelhttp/transport.go
Comment (MadVikingGod) [reply instrumentation/net/http/otelhttp/transport.go:165]: I'm saying that if we don't do anything in the record method, we shouldn't wrap it.  
That is to allow us to record a metric of how much was read.

File: instrumentation/net/http/otelhttp/common.go
Comment (MadVikingGod) [reply instrumentation/net/http/otelhttp/common.go:36]: Yes, we can make breaking changes, but no, we shouldn't do this lightly.

We need to deprecate these constants before we pull them out.  Pulling them out can make some user's code incompatible with newer versions of this package (and potentially the broader ecosystem).  While I do think that there is a low chance this will break any of the [827](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp) that import this library, we should act carefully around breaking changes.

File: instrumentation/net/http/otelhttp/transport.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go:165]: apologies, you're absolutely right. this is some left over cruft from the old PR where it was being used to record events. Will remove this shortly.

Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go:165]: We aren't doing anything in the record _callback_, however we are still using the wrapper to pull the number of bytes that were read.

File: instrumentation/net/http/otelhttp/common.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/common.go]: Added back and added docs about deprecation here: 70c1791e8dbdcd7324c5b1c78645b3e4a2b1d695

**Review by MadVikingGod — APPROVED** [review]

**Comment by Sovietaced** [comment]
@pellared should be good to go

File: instrumentation/net/http/otelhttp/transport_test.go
Comment (pellared) [inline instrumentation/net/http/otelhttp/transport_test.go:297]:
```suggestion
	assert.True(t, called, "record should have been called")
```

File: instrumentation/net/http/otelhttp/transport_test.go
Comment (pellared) [inline instrumentation/net/http/otelhttp/transport_test.go:315]:
```suggestion
	assert.True(t, called, "record should have been called")
```

File: instrumentation/net/http/otelhttp/transport.go
Comment (pellared) [inline instrumentation/net/http/otelhttp/transport.go:269]: Maybe add a comment here that we do not need to use `sync.Once` as it is OK that have "eventual consistency"? Basically if there are two goroutines running in parallel that call `recordBytesRead`, one can return before the second called `wb.record(wb.read.Load())`. It is OK and using `CompareAndSwap` instead of `sync.Once` is a performance optimization, but I think it is worth adding a comment.

**Review by pellared — COMMENTED** [review]
Little comments that I would like to have addressed. Besides looks good 👍

File: instrumentation/net/http/otelhttp/transport.go
Comment (Sovietaced) [reply instrumentation/net/http/otelhttp/transport.go:273]: Interesting, I'm relatively new to go so I had not known about `sync.Once`. I just wanted semantics close to Java's atomic boolean so that is why I chose `atomic.bool`. Anyway, added comments in 38861442146dafb02345a3cd06d6bd96030c77f6

**Comment by gaiaz-iusipov** [comment]
It would be nice to have something similar to `otelhttp.WithRouteTag` for the transport, thus allowing the separation of different requests in metrics. For example: `otelhttp.WithRequestTag`.

**Comment by pellared** [comment]
@Sovietaced Thank you for your contribution

**Comment by Sovietaced** [comment]
> It would be nice to have something similar to `otelhttp.WithRouteTag` for the transport, thus allowing the separation of different requests in metrics. For example: `otelhttp.WithRequestTag`.

Maybe create a separate issue for this?

**Comment by StarpTech** [comment]
Hi, it would be great if new features like metrics are built behind a feature flag. For example, we use the middleware for everything expect metrics. As a workaround, we pass a Noop meter.

File: instrumentation/net/http/otelhttp/test/handler_test.go
Comment (StarpTech) [inline instrumentation/net/http/otelhttp/test/handler_test.go:53]: This looks to me like a breaking change. Why wasn't it communicated as such?

Comment (dashpole) [reply instrumentation/net/http/otelhttp/test/handler_test.go:53]: @SharpTech sorry about that.  I've opened https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4898 to add it to the changelog.

Comment (StarpTech) [reply instrumentation/net/http/otelhttp/test/handler_test.go:53]: Hi @dashpole wouldn't it be better to rollback so the deprecation statement remain true?

Comment (dashpole) [reply instrumentation/net/http/otelhttp/test/handler_test.go:53]: The deprecation statements refer to these constants: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4707/files#diff-be9e2412d1919396470e196d2be4bddf6997059892167fd8381a9cfa478933faR34

> I also think breaking stuff with a feature release isn't the right approach.

The otelhttp module is not stable, and is still released under a v0.* version, so breaking changes are expected.  You can see the versioning policy here: https://github.com/open-telemetry/opentelemetry-go/blob/main/VERSIONING.md.  The telemetry in otelhttp follows the semantic conventions for http metrics: https://opentelemetry.io/docs/specs/semconv/http/http-metrics/, so breaking changes to conventions require us to make breaking changes here.

Comment (StarpTech) [reply instrumentation/net/http/otelhttp/test/handler_test.go:53]: I understand, but users like me running this in production. We need to deprecate fields before we can upgrade. If I would want to use the new metrics I couldn't. I'm saying the feature and the breaking change could have been shipped in two PR  (releases).

**Pull request merged**
